### DISCUSSION TITLE
Config Parser Error after upgrading from V7 to V8

### DISCUSSION THREAD

**Discussion by KierenLWoods** [webdriverio/webdriverio#10450] [Q&A]
Good Morning,

I have recently updated the following dependencies in a test automation framework from V7 to V8:
 "@wdio/allure-reporter": "^8.8.2",
    "@wdio/cli": "^8.10.5",
    "@wdio/jasmine-framework": "^8.10.5",
    "@wdio/junit-reporter": "^8.8.0",
    "@wdio/local-runner": "^8.8.2",
    "@wdio/spec-reporter": "^8.8.0",
    "@wdio/types": "^8.10.4",
 "wdio-chromedriver-service": "^8.1.1"

however when I now try to run even a spec file I get an Error from the config parser as below:
' ERROR @wdio/config:ConfigParser: Failed loading configuration file: file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/wdio.conf.ts: Package subpath './build/constants' is not defined by "exports" in C:\Users\kieren.woods\Repos\BH4SF\talent-rover-sfdx\e2e-tests\node_modules\webdriverio\package.json
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './build/constants' is not defined by "exports" in C:\Users\kieren.woods\Repos\BH4SF\talent-rover-sfdx\e2e-tests\node_modules\webdriverio\package.json
    at new NodeError (node:internal/errors:399:5)
    at exportsNotFound (node:internal/modules/esm/resolve:361:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:697:9)
    at resolveExports (node:internal/modules/cjs/loader:565:36)
    at Function.Module._findPath (node:internal/modules/cjs/loader:634:31)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1061:27)
    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (C:\Users\kieren.woods\Repos\BH4SF\talent-rover-sfdx\e2e-tests\node_modules\@cspotcode\source-map-support\source-map-support.js:811:30)
    at Function.Module._load (node:internal/modules/cjs/loader:920:27)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18) {
  code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'
}'

I didn't want to raise a bug for this as i believe it may just need some tweaks to my wdio.conf file and my tsConfig file but i have not been able to resolve this, just wanted to reach out to see if anyone else has encountered this.

This is my wdio.conf.ts file :
import type { Options } from '@wdio/types'
import { UtamWdioService } from 'wdio-utam-service';
import path from 'path';

// use this remote variable to distingusih decide local vs remote exections
global.remote = process.argv.some(arg => arg === '--remote');

export const config: Options.Testrunner = {
    //
    // ====================
    // Runner Configuration
    // ====================
    //
    //
    // =====================
    // ts-node Configurations
    // =====================
    //
    // You can write tests using TypeScript to get autocompletion and type safety.
    // You will need typescript and ts-node installed as devDependencies.
    // WebdriverIO will automatically detect if these dependencies are installed
    // and will compile your config and tests for you.
    // If you need to configure how ts-node runs please use the
    // environment variables for ts-node or use wdio config's autoCompileOpts section.
    //

    autoCompileOpts: {
        autoCompile: true,
        // see https://github.com/TypeStrong/ts-node#cli-and-programmatic-options
        // for all available options
        tsNodeOpts: {
            transpileOnly: true,
            project: 'tsconfig.json'
        },
        // tsconfig-paths is only used if "tsConfigPathsOpts" are provided, if you
        // do please make sure "tsconfig-paths" is installed as dependency
        tsConfigPathsOpts: {
            baseUrl: './',
            paths: {
                "@bh-domain/*": [
                    "domain/*"
                ],
                "@bh-pageObjects/*": [
                    "pageObjects/*"
                ],
                "@bh-pageUtils/*": [
                    "pageUtils/*"
                ],
                "@bh-utils/*": [
                    "utils/*"
                ]
            }
        }
    },
    //
    // ==================
    // Specify Test Files
    // ==================
    // Define which test specs should run. The pattern is relative to the directory
    // from which `wdio` was called.
    //
    // The specs are defined as an array of spec files (optionally using wildcards
    // that will be expanded). The test for each spec file will be run in a separate
    // worker process. In order to have a group of spec files run in the same worker
    // process simply enclose them in an array within the specs array.
    //
    // If you are calling `wdio` from an NPM script (see https://docs.npmjs.com/cli/run-script),
    // then the current working directory is where your `package.json` resides, so `wdio`
    // will be called from there.
    //
    runner: 'local',
    protocol: 'http',
    hostname: 'xx.xx.xx.xx',
    port: 1111,
    path: '/wd/hub',
    specs: [],
    suites: {

        
    },

    // Patterns to exclude.
    exclude: [
        // 'path/to/excluded/files'
    ],
    //
    // ============
    // Capabilities
    // ============
    // Define your capabilities here. WebdriverIO can run multiple capabilities at the same
    // time. Depending on the number of capabilities, WebdriverIO launches several test
    // sessions. Within your capabilities you can overwrite the spec and exclude options in
    // order to group specific specs to a specific capability.
    //
    // First, you can define how many instances should be started at the same time. Let's
    // say you have 3 different capabilities (Chrome, Firefox, and Safari) and you have
    // set maxInstances to 1; wdio will spawn 3 processes. Therefore, if you have 10 spec
    // files and you set maxInstances to 10, all spec files will get tested at the same time
    // and 30 processes will get spawned. The property handles how many capabilities
    // from the same test should run tests.
    //
    maxInstances: 5,
    //
    // If you have trouble getting all important capabilities together, check out the
    // Sauce Labs platform configurator - a great tool to configure your capabilities:
    // https://saucelabs.com/platform/platform-configurator
    //
    capabilities: [{

        // maxInstances can get overwritten per capability. So if you have an in-house Selenium
        // grid with only 5 firefox instances available you can make sure that not more than
        // 5 instances get started at a time.
        //maxInstances: 5,
        //
        browserName: 'chrome',
        acceptInsecureCerts: true,
        'selenoid:options': {
            "enableVNC": true,
            "sessionTimeout": "4m"
        },
        'goog:chromeOptions': {
            args: [
                'start-maximized',
                'disable-extensions',
                'disable-popup-blocking',
                'disable-notifications',
                'disable-login-animations'
            ],
            prefs: {
                'profile.default_content_settings.popups': 0,
                'directory_upgrade': true,
                'disable-popup-blocking': true,
                'download.prompt_for_download': false,
                'profile.default_content_setting_values.automatic_downloads': 1,
                'plugins.always_open_pdf_externally': true
            }
        },
        // If outputDir is provided WebdriverIO can capture driver session logs
        // it is possible to configure which logTypes to include/exclude.
        // excludeDriverLogs: ['*'], // pass '*' to exclude all driver session logs
        // excludeDriverLogs: ['bugreport', 'server'],
    }],
    //
    // ===================
    // Test Configurations
    // ===================
    // Define all options that are relevant for the WebdriverIO instance here
    //
    // Level of logging verbosity: trace | debug | info | warn | error | silent
    logLevel: 'error',
    //
    // Set specific log levels per logger
    // loggers:
    // - webdriver, webdriverio
    // - @wdio/browserstack-service, @wdio/devtools-service, @wdio/sauce-service
    // - @wdio/mocha-framework, @wdio/jasmine-framework
    // - @wdio/local-runner
    // - @wdio/sumologic-reporter
    // - @wdio/cli, @wdio/config, @wdio/utils
    // Level of logging verbosity: trace | debug | info | warn | error | silent
    // logLevels: {
    //     webdriver: 'info',
    //     '@wdio/appium-service': 'info'
    // },
    //
    // If you only want to run your tests until a specific amount of tests have failed use
    // bail (default is 0 - don't bail, run all tests).
    bail: 0,
    //
    // Set a base URL in order to shorten url command calls. If your `url` parameter starts
    // with `/`, the base url gets prepended, not including the path portion of your baseUrl.
    // If your `url` parameter starts without a scheme or `/` (like `some/path`), the base url
    // gets prepended directly.
    baseUrl: 'http://localhost',
    //
    // Default timeout for all waitFor* commands.
    waitforTimeout: 45_000,
    //
    // Default timeout in milliseconds for request
    // if browser driver or grid doesn't send response
    connectionRetryTimeout: 120_000,
    //
    // Default request retries count
    connectionRetryCount: 3,
    //
    // Test runner services
    // Services take over a specific job you don't want to take care of. They enhance
    // your test setup with almost no effort. Unlike plugins, they don't add new
    // commands. Instead, they hook themselves up into the test process.
    services: [...(global.remote ? [] : ['chromedriver']), [UtamWdioService, { implicitTimeout: 0, injectionConfigs: ['salesforce-pageobjects/ui-utam-pageobjects.config.json'] }]],

    // Framework you want to run your specs with.
    // The following are supported: Mocha, Jasmine, and Cucumber
    // see also: https://webdriver.io/docs/frameworks
    //
    // Make sure you have the wdio adapter package for the specific framework installed
    // before running any tests.
    framework: 'jasmine',
    //
    // The number of times to retry the entire specfile when it fails as a whole
    // specFileRetries: 1,
    //
    // Delay in seconds between the spec file retry attempts
    // specFileRetriesDelay: 0,
    //
    // Whether or not retried specfiles should be retried immediately or deferred to the end of the queue
    // specFileRetriesDeferred: false,
    //
    // Test reporter for stdout.
    // The only one supported by default is 'dot'
    // see also: https://webdriver.io/docs/dot-reporter
    reporters: [
        'spec',
        ['allure', {
            outputDir: './../allure-results',
            disableWebdriverStepsReporting: true,
            disableWebdriverScreenshotsReporting: false,
            disableMochaHooks: true,
            tmsLinkTemplate: ""
        }],
        ['junit', {
            outputDir: './reports/junit',
            outputFileFormat: function (options: { cid: any; capabilities: { browserName: any; }; }) { // optional
                return `results-${options.cid}-${Date.now()}-${options.capabilities.browserName}.xml`
            },
            errorOptions: {
                error: 'message',
                failure: 'message',
                stacktrace: 'stack'
            }
        }]
    ],

    //
    // Options to be passed to Jasmine.
    jasmineOpts: {
        // Jasmine default timeout
        defaultTimeoutInterval: 120_000,
        // Cleans up stack trace and removes all traces of node module packages
        cleanStack: false,
        //
        // The Jasmine framework allows interception of each assertion in order to log the state of the application
        // or website depending on the result. For example, it is pretty handy to take a screenshot every time
        // an assertion fails.
        expectationResultHandler: async function (passed, assertion) {
            if (!passed) {
                try {
                    const filename = (assertion.error?.message || assertion.message).trim()
                        .replace(/[\r\n"<>:\/\\|?*]/g, ' ').slice(0, 50);
                    await browser.saveScreenshot(`./reports/assertionError_${filename}_${Date.now()}.png`);
                } catch (error) {
                    console.log(error);
                }
            }
        },
    },

    //
    // =====
    // Hooks
    // =====
    // WebdriverIO provides several hooks you can use to interfere with the test process in order to enhance
    // it and to build services around it. You can either apply a single function or an array of
    // methods to it. If one of them returns with a promise, WebdriverIO will wait until that promise got
    // resolved to continue.
    /**
     * Gets executed once before all workers get launched.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     */
    //  onPrepare: function (config, capabilities) {

    //  },
    /**
     * Gets executed before a worker process is spawned and can be used to initialise specific service
     * for that worker as well as modify runtime environments in an async fashion.
     * @param  {String} cid      capability id (e.g 0-0)
     * @param  {[type]} caps     object containing capabilities for session that will be spawn in the worker
     * @param  {[type]} specs    specs to be run in the worker process
     * @param  {[type]} args     object that will be merged with the main configuration once worker is initialized
     * @param  {[type]} execArgv list of string arguments passed to the worker process
     */
    // onWorkerStart: function (cid, caps, specs, args, execArgv) {
    // },
    /**
     * Gets executed just after a worker process has exited.
     * @param  {String} cid      capability id (e.g 0-0)
     * @param  {Number} exitCode 0 - success, 1 - fail
     * @param  {[type]} specs    specs to be run in the worker process
     * @param  {Number} retries  number of retries used
     */
    // onWorkerEnd: function (cid, exitCode, specs, retries) {
    // },
    /**
     * Gets executed just before initialising the webdriver session and test framework. It allows you
     * to manipulate configurations depending on the capability or spec.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that are to be run
     * @param {String} cid worker id (e.g. 0-0)
     */
    // beforeSession: function (config, capabilities, specs, cid) {
    // },
    /**
     * Gets executed before test execution begins. At this point you can access to all global
     * variables like `browser`. It is the perfect place to define custom commands.
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs        List of spec file paths that are to be run
     * @param {Object}         browser      instance of created browser/device session
     */
    // before: function (capabilities, specs) {
    // },
    /**
     * Runs before a WebdriverIO command gets executed.
     * @param {String} commandName hook command name
     * @param {Array} args arguments that command would receive
     */
    // beforeCommand: function (commandName, args) {
    // },
    /**
     * Hook that gets executed before the suite starts
     * @param {Object} suite suite details
     */
    // beforeSuite: function (suite) {
    // },
    /**
     * Function to be executed before a test (in Mocha/Jasmine) starts.
     */
    // beforeTest: function (test, context) {
    // },
    /**
     * Hook that gets executed _before_ a hook within the suite starts (e.g. runs before calling
     * beforeEach in Mocha)
     */
    // beforeHook: function (test, context) {
    // },
    /**
     * Hook that gets executed _after_ a hook within the suite starts (e.g. runs after calling
     * afterEach in Mocha)
     */
    // afterHook: function (test, context, { error, result, duration, passed, retries }) {
    // },
    /**
     * Function to be executed after a test (in Mocha/Jasmine only)
     * @param {Object}  test             test object
     * @param {Object}  context          scope object the test was executed with
     * @param {Error}   result.error     error object in case the test fails, otherwise `undefined`
     * @param {Any}     result.result    return object of test function
     * @param {Number}  result.duration  duration of test
     * @param {Boolean} result.passed    true if test has passed, otherwise false
     * @param {Object}  result.retries   informations to spec related retries, e.g. `{ attempts: 0, limit: 0 }`
     */
    // afterTest: function(test, context, { error, result, duration, passed, retries }) {
    // },

    /**
     * Hook that gets executed after the suite has ended
     * @param {Object} suite suite details
     */
    // afterSuite: function (suite) {
    // },
    /**
     * Runs after a WebdriverIO command gets executed
     * @param {String} commandName hook command name
     * @param {Array} args arguments that command would receive
     * @param {Number} result 0 - command success, 1 - command error
     * @param {Object} error error object if any
     */
    // afterCommand: function (commandName, args, result, error) {
    // },
    /**
     * Gets executed after all tests are done. You still have access to all global variables from
     * the test.
     * @param {Number} result 0 - test pass, 1 - test fail
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that ran
     */
    // after: function (result, capabilities, specs) {
    // },
    /**
     * Gets executed right after terminating the webdriver session.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that ran
     */
    // afterSession: function (config, capabilities, specs) {
    // },
    /**
     * Gets executed after all workers got shut down and the process is about to exit. An error
     * thrown in the onComplete hook will result in the test run failing.
     * @param {Object} exitCode 0 - success, 1 - fail
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {<Object>} results object containing test results
     */
    //  onComplete: function(exitCode, config, capabilities, results) {
    //  },
    /**
    * Gets executed when a refresh happens.
    * @param {String} oldSessionId session ID of the old session
    * @param {String} newSessionId session ID of the new session
    */
    // onReload: function(oldSessionId, newSessionId) {
    // }
}

and here is my tsConfig file:
{
    "compilerOptions": {
        "moduleResolution": "node",
        "target": "es2019",
        "esModuleInterop": true,
        "baseUrl": "./",
        "types": [
            "node",
            "webdriverio/async",
            "@wdio/jasmine-framework",
            "expect-webdriverio",
            "wdio-utam-service",
        ],
        "paths": {
            "@bh-domain/*": [
                "domain/*"
            ],
            "@bh-pageObjects/*": [
                "pageObjects/*"
            ],
            "@bh-utils/*": [
                "utils/*"
            ],
            "@bh-pageUtils/*": [
                "pageUtils/*"
            ],
            "salesforce-pageobjects/*": [
                "node_modules/salesforce-pageobjects/dist/*"
            ],
        }
    }
}

Any help would be greatly appreciated.

**Comment by christian-bromann** [comment]
> ' ERROR @wdio/config:ConfigParser: Failed loading configuration file: file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/wdio.conf.ts: Package subpath './build/constants' is not defined by "exports" in C:\Users\kieren.woods\Repos\BH4SF\talent-rover-sfdx\e2e-tests\node_modules\webdriverio\package.json
> Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './build/constants' is not defined by "exports" in C:\Users\kieren.woods\Repos\BH4SF\talent-rover-sfdx\e2e-tests\node_modules\webdriverio\package.json

This error sounds to me that not all dependencies are correctly updated to v8

**Reply by KierenLWoods** [reply]
@christian-bromann thank you for your response. I've taken a look in the dependencies and within the wdio-utam-service dependency this still references  "@wdio/logger": "^7.26.0" and "@wdio/types": "^7.30.2" could that be whats causing the issue?

**Reply by christian-bromann** [reply]
yes, try to load services like this:

```
services: [
    ...(global.remote ? [] : ['chromedriver']),
    ['utam', {
        implicitTimeout: 0,
        injectionConfigs: ['salesforce-pageobjects/ui-utam-pageobjects.config.json']
    }
]],
```

so WebdriverIO can load them in a way that it would support CJS environments and v7 services.

**Reply by KierenLWoods** [reply]
@christian-bromann thank you again for getting back to me. I made those changes to the wdio file however i am now getting this error when trying to run a test:
Error in "undefined"
Error: Failed to initilialise launcher service unknown: Error: Couldn't initialise "wdio-utam-service".
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './build/constants' is not defined by "exports" in C:\Users\kieren.woods\Repos\BH4SF\talent-rover-sfdx\e2e-tests\node_modules\webdriverio\package.json
    at new NodeError (node:internal/errors:393:5)
    at throwExportsNotFound (node:internal/modules/esm/resolve:358:9)
    at packageExportsResolve (node:internal/modules/esm/resolve:668:3)
    at resolveExports (node:internal/modules/cjs/loader:529:36)
    at Function.Module._findPath (node:internal/modules/cjs/loader:569:31)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:981:27)
    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (C:\Users\kieren.woods\Repos\BH4SF\talent-rover-sfdx\e2e-tests\node_modules\@cspotcode\source-map-support\source-map-support.js:811:30)
    at Function.Module._load (node:internal/modules/cjs/loader:841:27)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at safeImport (file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/node_modules/@wdio/utils/build/utils.js:210:15)
    at async initialisePlugin (file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/node_modules/@wdio/utils/build/initialisePlugin.js:32:20)
    at async initialiseServices (file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/node_modules/@wdio/utils/build/initialiseServices.js:53:25)
    at async initialiseLauncherService (file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/node_modules/@wdio/utils/build/initialiseServices.js:81:26)
    at async Launcher.run (file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/node_modules/@wdio/cli/build/launcher.js:69:65)
    at initialiseLauncherService (file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/node_modules/@wdio/utils/build/initialiseServices.js:118:15)
    at async Launcher.run (file:///C:/Users/kieren.woods/Repos/BH4SF/talent-rover-sfdx/e2e-tests/node_modules/@wdio/cli/build/launcher.js:69:65)

**Reply by christian-bromann** [reply]
@KierenLWoods can you create a minimal reproducible example in a new repository?

**Reply by KierenLWoods** [reply]
@christian-bromann  heres a link to a reproducible example i have created https://github.com/KierenLWoods/webDriverIOIssue.git, when i run the spec file against v7 dependencies the test runs and fails as expected. However when i update dependencies to V8 i then run into the issue above.

**Reply by KierenLWoods** [reply]
@christian-bromann just wondered if you were able to replicate this with the repo i sent?

**Reply by christian-bromann** [reply]
sorry but I don't have much time to look into this

**Reply by Durga909** [reply]
> @christian-bromann heres a link to a reproducible example i have created https://github.com/KierenLWoods/webDriverIOIssue.git, when i run the spec file against v7 dependencies the test runs and fails as expected. However when i update dependencies to V8 i then run into the issue above.

Hi @KierenLWoods 
I do face the same issues with UTAM on V8. Do we have any resolution at this point.

**Reply by olivier-martin-sf** [reply]
@Durga909 Version 2.2.0 of UTAM added support for v8; prior to that, only v7 was compatible. If you are a UTAM user, we have a dedicated Slack channel where you can directly ask our team for support. If you would like to be added to the channel, please reach out to me on Twitter (@anchnk).

**Comment by JiangHongTiao** [comment]
I've experienced similar problem when I did
`import {sleep} from "@wdio/browserstack-service/build/util";`

During execution I've got: 
`@wdio/runner: Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './build/util' is not defined by "exports" in ./node_modules/@wdio/browserstack-service/package.json`
